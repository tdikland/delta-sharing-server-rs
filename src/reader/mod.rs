//! Types and traits for reading table data in object storage.

use std::{collections::HashMap, error::Error, fmt::Display};

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

// TODO: refer to the official kernel types (not yet available)
use deltalake::kernel::{Add, AddCDCFile, Metadata, Protocol, Remove};

/// Table reader implementation for the Delta Lake format.
pub mod delta;

/// Trait for reading a specific table format from cloud storage.
#[cfg_attr(test, mockall::automock)]
#[async_trait]
pub trait TableReader: Send + Sync {
    /// Retrieve the table version number that corresponds to the version
    /// request.
    async fn get_table_version_number(
        &self,
        storage_path: &str,
        version: Version,
    ) -> Result<TableVersionNumber, TableReaderError>;

    /// Retrieve the table metadata corresponding to the latest table version.
    async fn get_table_metadata(
        &self,
        storage_path: &str,
    ) -> Result<TableMetadata, TableReaderError>;

    /// Retrieve the table data for a specific table version.
    ///
    /// The table data is represented by a collection of files which can be
    /// directly reached with a presigned HTTPS url. The limit and predicate
    /// argument can be used to restrict the returned data files on a best
    /// effort basis.
    async fn get_table_data(
        &self,
        storage_path: &str,
        version: Version,
        limit: Option<u64>,
        predicates: Option<String>,
        opt: Option<HashMap<String, String>>,
    ) -> Result<TableData, TableReaderError>;

    /// Retrieve the table change data for a specific range of table versions.
    ///
    /// The table changes are represented by a collection of files which can be
    /// directly reached with presigned HTTPS urls.
    async fn get_table_changes(
        &self,
        storage_path: &str,
        range: VersionRange,
        opt: Option<HashMap<String, String>>,
    ) -> Result<TableData, TableReaderError>;
}

/// Table version number.
pub struct TableVersionNumber(u64);

/// Table metadata for a given table version.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TableMetadata {
    version: u64,
    protocol: Protocol,
    metadata: Metadata,
    metadata_num_files: Option<u64>,
    metadata_size: Option<u64>,
}

/// Table metadata and data descriptors, not yet publicly accessible.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TableData {
    version: u64,
    protocol: Protocol,
    metadata: Metadata,
    metadata_num_files: Option<u64>,
    metadata_size: Option<u64>,
    data: Vec<(Action, Option<u64>)>,
}

pub enum Action {
    Protocol(Protocol),
    Metadata(Metadata),
    Add(Add),
    Cdf(AddCDCFile),
    Remove(Remove),
}

/// A representation of data or mutation in a table referenced using an object
/// store url.
///
/// A table is represented as a set of files that together are the full table.
/// Every data file has a reference to the underlying object store in its url
/// field.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum UnsignedDataFile {
    /// A file containing data part of the table.
    File(Add),
    /// A file containing data that was added to the table in this version.
    Add(Add),
    /// A file containing data that was changed in this version of the table.
    Cdf(AddCDCFile),
    /// A file containing data that was removed since the last table version.
    Remove(Remove),
}

/// Requested table version.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Version {
    /// Latest table version.
    Latest,
    /// Table version number.
    Number(u64),
    /// Earliest table version after the specified timestamp.
    Timestamp(DateTime<Utc>),
}

/// Requested range of table version.
#[derive(Debug, Clone, Copy)]
pub enum VersionRange {
    /// Range of table versions represented by start and end version number.
    Version {
        /// First timestamp that must be returned in the range.
        start: u64,
        /// Last timestamp that must be returned in the range.
        end: u64,
    },
    /// Range of table versions represented by start and end timestamp.
    Timestamp {
        /// First version must be the earliest after the start timestamp.
        start: DateTime<Utc>,
        /// Last version must be the earliest after the end timestamp.
        end: DateTime<Utc>,
    },
}

/// Error that occur during the reading of the table format.
#[derive(Debug, Clone)]
pub enum TableReaderError {
    /// An unexpected error occured.
    Other,
}

impl Display for TableReaderError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TableReaderError::Other => {
                write!(f, "An unexpected error happened during table reading")
            }
        }
    }
}

impl Error for TableReaderError {}
