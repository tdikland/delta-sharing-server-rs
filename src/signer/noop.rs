//! A signer that does not sign URLs.

use async_trait::async_trait;

use super::{SignedUrl, UrlSigner};

/// A signer that does not sign URLs.
#[derive(Debug)]
pub struct NoopSigner;

#[async_trait]
impl UrlSigner for NoopSigner {
    async fn sign_url(&self, path: &str) -> SignedUrl {
        SignedUrl::new(
            path.to_string(),
            chrono::Utc::now(),
            std::time::Duration::from_secs(3600),
        )
    }
}
