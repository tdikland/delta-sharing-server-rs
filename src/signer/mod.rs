//! Traits and types for creating pre-signed urls.

use std::time::Duration;

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::reader::UnsignedDataFile;
use deltalake::kernel::{Add, AddCDCFile, Metadata, Protocol, Remove};

pub mod registry;

mod adls;
mod gcs;
pub mod noop;
pub mod s3;

/// Trait implemented by object store clients to derive a pre-signed url from
/// a object store path/prefix.
#[cfg_attr(test, mockall::automock)]
#[async_trait]
pub trait UrlSigner: Send + Sync {
    /// Create a presigned url from a object store path.
    async fn sign_url(&self, path: &str) -> SignedUrl;

    /// Create a presigned url for a object store path within a data file.
    async fn sign_data_file(&self, data_file: crate::reader::UnsignedDataFile) -> SignedDataFile {
        match data_file {
            UnsignedDataFile::File(mut file) => {
                file.path = self.sign_url(&file.path).await.url;
                SignedDataFile::File(file)
            }
            UnsignedDataFile::Add(mut add) => {
                add.path = self.sign_url(&add.path).await.url;
                SignedDataFile::Add(add)
            }
            UnsignedDataFile::Cdf(mut cdf) => {
                cdf.path = self.sign_url(&cdf.path).await.url;
                SignedDataFile::Cdf(cdf)
            }
            UnsignedDataFile::Remove(mut remove) => {
                remove.path = self.sign_url(&remove.path).await.url;
                SignedDataFile::Remove(remove)
            }
        }
    }

    /// Create presigned urls for all data files in a table version.
    async fn sign_table_data(
        &self,
        table_data: crate::reader::UnsignedTableData,
    ) -> SignedTableData {
        let mut signed_data_files = vec![];
        for data_file in table_data.data {
            signed_data_files.push(self.sign_data_file(data_file).await);
        }
        SignedTableData {
            version: table_data.version,
            protocol: table_data.protocol,
            metadata: table_data.metadata,
            data: signed_data_files,
        }
    }
}

/// A presigned url with a validity period.
pub struct SignedUrl {
    url: String,
    valid_from: DateTime<Utc>,
    valid_duration: Duration,
}

impl SignedUrl {
    fn new(url: String, valid_from: DateTime<Utc>, valid_duration: Duration) -> Self {
        Self {
            url,
            valid_from,
            valid_duration,
        }
    }

    /// Get the presigned url.
    pub fn url(&self) -> &str {
        &self.url
    }

    /// Get the time the presigned url expires.
    pub fn expires_at(&self) -> DateTime<Utc> {
        self.valid_from + self.valid_duration
    }
}

/// Table metadata and data descriptors with presigned urls.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SignedTableData {
    /// Table version.
    pub version: u64,
    /// Minimum required table reader protocol implementation.
    pub protocol: Protocol,
    /// Table metadata
    pub metadata: Metadata,
    /// Set of data file representing the table
    pub data: Vec<SignedDataFile>,
}

/// A representation of data or mutation in a table reachable with a presigned
/// url.
///
/// A table is represented as a set of files that together are the full table.
/// Every data file has a presigned url that can be used to directly access the
/// data file.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
#[serde(untagged)]
pub enum SignedDataFile {
    /// A file containing data part of the table.
    File(Add),
    /// A file containing data that was added to the table in this version.
    Add(Add),
    /// A file containing data that was changed in this version of the table.
    Cdf(AddCDCFile),
    /// A file containing data that was removed since the last table version.
    Remove(Remove),
}
