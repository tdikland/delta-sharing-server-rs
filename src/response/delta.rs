use std::{io::Write, sync::Arc};

use axum::response::{IntoResponse, Response};
use bytes::{BufMut, BytesMut};
use delta_kernel::actions::{Add, Metadata, Protocol};
use http::{header, StatusCode};
use serde::Serialize;
use url::Url;

use crate::reader::{TableData, TableMeta};
use crate::signer::UrlSigner;

pub struct DeltaResponse {
    version: u64,
    protocol: DeltaResponseLine,
    metadata: DeltaResponseLine,
    lines: Vec<DeltaResponseLine>,
}

impl DeltaResponse {
    pub async fn sign(&mut self, table_root: &str, signer: Arc<dyn UrlSigner>) {
        for line in self.lines.iter_mut() {
            line.sign(table_root, signer.clone()).await;
        }
    }
}

impl IntoResponse for DeltaResponse {
    fn into_response(self) -> Response {
        let mut buf = BytesMut::new().writer();

        serde_json::to_writer(&mut buf, &self.protocol).unwrap();
        buf.write_all(b"\n").unwrap();

        serde_json::to_writer(&mut buf, &self.metadata).unwrap();
        buf.write_all(b"\n").unwrap();

        for line in self.lines {
            serde_json::to_writer(&mut buf, &line).unwrap();
            buf.write_all(b"\n").unwrap();
        }

        let version = self.version.to_string();
        let headers = [
            (
                header::CONTENT_TYPE.as_str(),
                "application/x-ndjson; charset=utf-8",
            ),
            ("Delta-Table-Version", version.as_ref()),
        ];

        (StatusCode::OK, headers, buf.into_inner()).into_response()
    }
}

impl From<TableMeta> for DeltaResponse {
    fn from(meta: TableMeta) -> Self {
        let protocol = ProtocolResponseLine::from_protocol(meta.protocol().clone().into_inner());
        let metadata = MetadataResponseLine::from_metadata_with_opts(
            meta.metadata().clone().into_inner(),
            Some(meta.version()),
            None,
            None,
        );

        Self {
            version: meta.version(),
            protocol: protocol.into(),
            metadata: metadata.into(),
            lines: vec![],
        }
    }
}

impl From<TableData> for DeltaResponse {
    fn from(data: TableData) -> Self {
        let protocol = ProtocolResponseLine::from_protocol(data.protocol().clone().into_inner());
        let metadata = MetadataResponseLine::from_metadata_with_opts(
            data.metadata().clone().into_inner(),
            Some(data.version()),
            None,
            None,
        );

        let mut lines = vec![];
        for file in data.data().into_iter().cloned() {
            lines.push(
                FileResponseLine::from_add_with_opts(file.into_inner(), None, None, None).into(),
            );
        }
        Self {
            version: data.version(),
            protocol: protocol.into(),
            metadata: metadata.into(),
            lines,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
enum DeltaResponseLine {
    Protocol(ProtocolResponseLine),
    #[serde(rename = "metaData")]
    Metadata(MetadataResponseLine),
    File(FileResponseLine),
}

impl DeltaResponseLine {
    async fn sign(&mut self, table_root: &str, signer: Arc<dyn UrlSigner>) {
        match self {
            Self::File(file) => {
                file.sign(table_root, signer).await;
            }
            _ => {}
        }
    }
}

impl From<ProtocolResponseLine> for DeltaResponseLine {
    fn from(line: ProtocolResponseLine) -> Self {
        Self::Protocol(line)
    }
}

impl From<MetadataResponseLine> for DeltaResponseLine {
    fn from(line: MetadataResponseLine) -> Self {
        Self::Metadata(line)
    }
}

impl From<FileResponseLine> for DeltaResponseLine {
    fn from(line: FileResponseLine) -> Self {
        Self::File(line)
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct ProtocolResponseLine {
    delta_protocol: Protocol,
}

impl ProtocolResponseLine {
    fn from_protocol(protocol: Protocol) -> Self {
        Self {
            delta_protocol: protocol,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct MetadataResponseLine {
    delta_metadata: Metadata,
    version: Option<u64>,
    size: Option<u64>,
    num_files: Option<u64>,
}

impl MetadataResponseLine {
    fn from_metadata_with_opts(
        metadata: Metadata,
        version: Option<u64>,
        size: Option<u64>,
        num_files: Option<u64>,
    ) -> Self {
        Self {
            delta_metadata: metadata,
            version,
            size,
            num_files,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
enum SingleAction {
    Add(Add),
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct FileResponseLine {
    id: String,
    deletion_vector_file_id: Option<String>,
    version: Option<u64>,
    timestamp: Option<u64>,
    expiration_timestamp: Option<u64>,
    delta_single_action: SingleAction,
}

impl FileResponseLine {
    fn from_add_with_opts(
        add: Add,
        version: Option<u64>,
        timestamp: Option<u64>,
        expiration_timestamp: Option<u64>,
    ) -> Self {
        let file_id = format!("{:x}", md5::compute(add.path.as_bytes()));
        // TODO: derive deletion_vector_file_id
        Self {
            id: file_id,
            deletion_vector_file_id: None,
            version,
            timestamp,
            expiration_timestamp,
            delta_single_action: SingleAction::Add(add),
        }
    }

    async fn sign(&mut self, table_root: &str, signer: Arc<dyn UrlSigner>) {
        match &mut self.delta_single_action {
            SingleAction::Add(add) => {
                let file_url = format!("{}/{}", table_root, add.path);
                let signed_url = signer.sign_url(&file_url).await;
                add.path = signed_url.url().to_string();

                if let Some(dv) = &mut add.deletion_vector {
                    match dv.storage_type.as_str() {
                        "i" => (),
                        "u" => {
                            let parent = Url::parse(table_root).unwrap();
                            let deletion_vector_url = dv.absolute_path(&parent).unwrap().unwrap();
                            let signed_deletion_vector_url =
                                signer.sign_url(deletion_vector_url.as_str()).await;

                            dv.storage_type = "p".to_string();
                            dv.path_or_inline_dv = signed_deletion_vector_url.url().to_string();
                        }
                        "p" => {
                            let deletion_vector_url = dv.path_or_inline_dv.clone();
                            let signed_dv = signer.sign_url(&deletion_vector_url).await;

                            dv.path_or_inline_dv = signed_dv.url().to_string();
                        }
                        _ => {}
                    }
                }
            }
        }
    }
}

#[cfg(test)]
mod test {
    use std::{collections::HashMap, time::Duration};

    use chrono::{TimeZone, Utc};
    use delta_kernel::actions::DeletionVectorDescriptor;
    use delta_kernel::actions::Format;
    use insta::assert_json_snapshot;

    use crate::signer::{MockUrlSigner, SignedUrl};

    use super::*;

    #[test]
    fn build_delta_protocol_line() {
        let protocol = Protocol {
            min_reader_version: 1,
            min_writer_version: 2,
            reader_features: None,
            writer_features: None,
        };

        let protocol_line = ProtocolResponseLine::from_protocol(protocol);
        let line = DeltaResponseLine::Protocol(protocol_line);

        assert_json_snapshot!(line);
    }

    #[test]
    fn build_delta_metadata_line() {
        let metadata = Metadata {
            id: String::from("f8d5c169-3d01-4ca3-ad9e-7dc3355aedb2"),
            name: Some(String::from("name")),
            description: Some(String::from("description")),
            format: Format {
                provider: String::from("parquet"),
                options: HashMap::new(),
            },
            schema_string: String::from("{\"type\":\"struct\",\"fields\":[{\"name\":\"eventTime\",\"type\":\"timestamp\",\"nullable\":true,\"metadata\":{}}]}"),
            partition_columns: vec![String::from("date")],
            created_time: Some(1619824428000),
            configuration: HashMap::from_iter([("opt1".to_owned(), Some("true".to_owned()))]),
        };

        let metadata_line =
            MetadataResponseLine::from_metadata_with_opts(metadata, Some(1), Some(2), Some(3));
        let line = DeltaResponseLine::Metadata(metadata_line);

        assert_json_snapshot!(line);
    }

    #[test]
    fn build_delta_file_line() {
        let add = Add {
            path: "https://bucket/key".to_owned(),
            partition_values: HashMap::from_iter([(
                "date".to_owned(),
                Some("2021-04-28".to_owned()),
            )]),
            size: 573,
            stats: Some("{\"numRecords\":1}".to_owned()),
            modification_time: 1619824428000,
            data_change: false,
            tags: HashMap::new(),
            deletion_vector: None,
            base_row_id: None,
            default_row_commit_version: None,
            clustering_provider: None,
        };

        let file_line = FileResponseLine::from_add_with_opts(add, Some(1), Some(2), Some(3));
        let line = DeltaResponseLine::File(file_line);

        assert_json_snapshot!(line);
    }

    #[tokio::test]
    async fn sign_delta_file_line_without_deletion_vector() {
        let mut mock_signer = MockUrlSigner::new();
        mock_signer.expect_sign_url().returning(|s| {
            let url = format!("{s}?sig=foo_signature");
            let valid_from = Utc.timestamp_opt(1_610_000_000, 0).unwrap();
            let expiration = Duration::from_secs(3600);
            SignedUrl::new(url, valid_from, expiration)
        });

        let add = Add {
            path: "key.snappy.parquet".to_owned(),
            partition_values: HashMap::new(),
            size: 573,
            stats: None,
            modification_time: 1619824428000,
            data_change: false,
            tags: HashMap::new(),
            deletion_vector: None,
            base_row_id: None,
            default_row_commit_version: None,
            clustering_provider: None,
        };

        let mut file_line = FileResponseLine::from_add_with_opts(add, None, None, None);
        let table_root = "https://bucket/prefix";
        let signer = Arc::new(mock_signer);
        file_line.sign(table_root, signer).await;

        assert_json_snapshot!(file_line);
    }

    #[tokio::test]
    async fn sign_delta_file_line_inline_deletion_vector() {
        let mut mock_signer = MockUrlSigner::new();
        mock_signer.expect_sign_url().returning(|s| {
            let url = format!("{s}?sig=foo_signature");
            let valid_from = Utc.timestamp_opt(1_610_000_000, 0).unwrap();
            let expiration = Duration::from_secs(3600);
            SignedUrl::new(url, valid_from, expiration)
        });

        let add = Add {
            path: "key.snappy.parquet".to_owned(),
            partition_values: HashMap::new(),
            size: 573,
            stats: None,
            modification_time: 1619824428000,
            data_change: false,
            tags: HashMap::new(),
            deletion_vector: Some(DeletionVectorDescriptor {
                storage_type: "i".to_string(),
                path_or_inline_dv: "wi5b=000010000siXQKl0rr91000f55c8Xg0@@D72lkbi5=-{L".to_string(),
                offset: None,
                size_in_bytes: 40,
                cardinality: 6,
            }),
            base_row_id: None,
            default_row_commit_version: None,
            clustering_provider: None,
        };

        let mut file_line = FileResponseLine::from_add_with_opts(add, None, None, None);
        let table_root = "https://bucket/prefix";
        let signer = Arc::new(mock_signer);
        file_line.sign(table_root, signer).await;

        assert_json_snapshot!(file_line);
    }

    #[tokio::test]
    async fn sign_delta_file_line_relative_deletion_vector() {
        let mut mock_signer = MockUrlSigner::new();
        mock_signer.expect_sign_url().returning(|s| {
            let url = format!("{s}?sig=foo_signature");
            let valid_from = Utc.timestamp_opt(1_610_000_000, 0).unwrap();
            let expiration = Duration::from_secs(3600);
            SignedUrl::new(url, valid_from, expiration)
        });

        let add = Add {
            path: "key.snappy.parquet".to_owned(),
            partition_values: HashMap::new(),
            size: 573,
            stats: None,
            modification_time: 1619824428000,
            data_change: false,
            tags: HashMap::new(),
            deletion_vector: Some(DeletionVectorDescriptor {
                storage_type: "u".to_string(),
                path_or_inline_dv: "ab^-aqEH.-t@S}K{vb[*k^".to_string(),
                offset: Some(4),
                size_in_bytes: 40,
                cardinality: 6,
            }),
            base_row_id: None,
            default_row_commit_version: None,
            clustering_provider: None,
        };

        let mut file_line = FileResponseLine::from_add_with_opts(add, None, None, None);
        let table_root = "https://bucket/prefix";
        let signer = Arc::new(mock_signer);
        file_line.sign(table_root, signer).await;

        assert_json_snapshot!(file_line);
    }

    #[tokio::test]
    async fn sign_delta_file_line_absolute_deletion_vector() {
        let mut mock_signer = MockUrlSigner::new();
        mock_signer.expect_sign_url().returning(|s| {
            let url = format!("{s}?sig=foo_signature");
            let valid_from = Utc.timestamp_opt(1_610_000_000, 0).unwrap();
            let expiration = Duration::from_secs(3600);
            SignedUrl::new(url, valid_from, expiration)
        });

        let add = Add {
            path: "key.snappy.parquet".to_owned(),
            partition_values: HashMap::new(),
            size: 573,
            stats: None,
            modification_time: 1619824428000,
            data_change: false,
            tags: HashMap::new(),
            deletion_vector: Some(DeletionVectorDescriptor {
                storage_type: "p".to_string(),
                path_or_inline_dv:
                    "s3://mytable/deletion_vector_d2c639aa-8816-431a-aaf6-d3fe2512ff61.bin"
                        .to_string(),
                offset: Some(4),
                size_in_bytes: 40,
                cardinality: 6,
            }),
            base_row_id: None,
            default_row_commit_version: None,
            clustering_provider: None,
        };

        let mut file_line = FileResponseLine::from_add_with_opts(add, None, None, None);
        let table_root = "https://bucket/prefix";
        let signer = Arc::new(mock_signer);
        file_line.sign(table_root, signer).await;

        assert_json_snapshot!(file_line);
    }
}
