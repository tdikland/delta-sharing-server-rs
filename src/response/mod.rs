use std::sync::Arc;

use axum::http::HeaderName;
use axum::response::Response;
use axum::{
    http::{header, StatusCode},
    response::IntoResponse,
    Json,
};
use serde::Serialize;

use self::delta::DeltaResponse;
use self::parquet::ParquetResponse;
use crate::catalog::{Page, Schema as SchemaInfo, Share as ShareInfo, Table as TableInfo};
use crate::reader::TableVersionNumber;
use crate::signer::UrlSigner;

pub mod delta;
pub mod parquet;

static DELTA_TABLE_VERSION: HeaderName = HeaderName::from_static("delta-table-version");

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Share {
    name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<String>,
}

impl From<ShareInfo> for Share {
    fn from(value: ShareInfo) -> Self {
        Self {
            name: value.name().to_owned(),
            id: value.id().map(|id| id.to_owned()),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSharesResponse {
    items: Vec<Share>,
    #[serde(skip_serializing_if = "Option::is_none")]
    next_page_token: Option<String>,
}

impl From<Page<ShareInfo>> for ListSharesResponse {
    fn from(value: Page<ShareInfo>) -> Self {
        let (items, next_page_token) = value.into_parts();

        Self {
            items: items.into_iter().map(Share::from).collect(),
            next_page_token,
        }
    }
}

impl IntoResponse for ListSharesResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GetShareResponse {
    share: Share,
}

impl From<ShareInfo> for GetShareResponse {
    fn from(value: ShareInfo) -> Self {
        Self {
            share: value.into(),
        }
    }
}

impl IntoResponse for GetShareResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    name: String,
    share: String,
}

impl From<SchemaInfo> for Schema {
    fn from(value: SchemaInfo) -> Self {
        Self {
            name: value.name().to_owned(),
            share: value.share_name().to_owned(),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSchemasResponse {
    items: Vec<Schema>,
    #[serde(skip_serializing_if = "Option::is_none")]
    next_page_token: Option<String>,
}

impl From<Page<SchemaInfo>> for ListSchemasResponse {
    fn from(value: Page<SchemaInfo>) -> Self {
        let (items, next_page_token) = value.into_parts();

        Self {
            items: items.into_iter().map(Schema::from).collect(),
            next_page_token,
        }
    }
}

impl IntoResponse for ListSchemasResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Table {
    name: String,
    schema: String,
    share: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    share_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<String>,
}

impl From<TableInfo> for Table {
    fn from(value: TableInfo) -> Self {
        Self {
            name: value.name().to_owned(),
            schema: value.schema_name().to_owned(),
            share: value.share_name().to_owned(),
            share_id: value.share_id().map(ToOwned::to_owned),
            id: value.id().to_owned().map(ToOwned::to_owned),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListTablesResponse {
    items: Vec<Table>,
    #[serde(skip_serializing_if = "Option::is_none")]
    next_page_token: Option<String>,
}

impl From<Page<TableInfo>> for ListTablesResponse {
    fn from(value: Page<TableInfo>) -> Self {
        let (items, next_page_token) = value.into_parts();

        Self {
            items: items.into_iter().map(Table::from).collect(),
            next_page_token,
        }
    }
}

impl IntoResponse for ListTablesResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct TableVersionResponse {
    version: TableVersionNumber,
}

impl From<TableVersionNumber> for TableVersionResponse {
    fn from(version: TableVersionNumber) -> Self {
        Self { version }
    }
}

impl IntoResponse for TableVersionResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(DELTA_TABLE_VERSION.clone(), self.version.version())],
        )
            .into_response()
    }
}

#[derive(Debug, Clone)]
pub enum TableActionsResponse {
    Parquet(ParquetResponse),
    Delta(DeltaResponse),
}

impl TableActionsResponse {
    pub fn new_parquet<R: Into<ParquetResponse>>(response: R) -> Self {
        Self::Parquet(response.into())
    }

    pub fn new_delta<R: Into<DeltaResponse>>(response: R) -> Self {
        Self::Delta(response.into())
    }

    pub async fn sign(self, table_root: &str, signer: Arc<dyn UrlSigner>) -> Self {
        match self {
            TableActionsResponse::Parquet(response) => {
                let signed_response = response.sign(table_root, signer).await;
                TableActionsResponse::Parquet(signed_response)
            }
            TableActionsResponse::Delta(mut response) => {
                response.sign(table_root, signer).await;
                TableActionsResponse::Delta(response)
            }
        }
    }
}

impl IntoResponse for TableActionsResponse {
    fn into_response(self) -> Response {
        match self {
            TableActionsResponse::Parquet(response) => response.into_response(),
            TableActionsResponse::Delta(response) => response.into_response(),
        }
    }
}

#[cfg(test)]
mod test {
    use axum::{
        body::to_bytes,
        http::{header::CONTENT_TYPE, HeaderValue},
    };
    use bytes::Bytes;

    use super::*;

    #[tokio::test]
    async fn list_shares_response() {
        let share_info = Page::new(
            vec![ShareInfo::builder()
                .name("share_name")
                .id("share_id")
                .build()
                .unwrap()],
            Some("page_token".to_owned()),
        );
        let response = ListSharesResponse::from(share_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(
                r#"{"items":[{"name":"share_name","id":"share_id"}],"nextPageToken":"page_token"}"#
            )
        );
    }

    #[tokio::test]
    async fn get_share_response() {
        let share_info = ShareInfo::builder()
            .name("share_name")
            .id("share_id")
            .build()
            .unwrap();
        let response = GetShareResponse::from(share_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(r#"{"share":{"name":"share_name","id":"share_id"}}"#)
        );
    }

    #[tokio::test]
    async fn list_schemas_response() {
        let schema_info = Page::new(
            vec![SchemaInfo::builder()
                .name("schema_name")
                .share_name("share_name")
                .build()
                .unwrap()],
            Some("page_token".to_owned()),
        );
        let response = ListSchemasResponse::from(schema_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(
                r#"{"items":[{"name":"schema_name","share":"share_name"}],"nextPageToken":"page_token"}"#
            )
        );
    }

    #[tokio::test]
    async fn list_tables_response() {
        let table_info = Page::new(
            vec![TableInfo::builder()
                .name("table_name")
                .schema_name("schema_name")
                .share_name("share_name")
                .id("table_id")
                .share_id("share_id")
                .storage_path("not important here")
                .build()
                .unwrap()],
            Some("page_token".to_owned()),
        );
        let response = ListTablesResponse::from(table_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(
                r#"{"items":[{"name":"table_name","schema":"schema_name","share":"share_name","shareId":"share_id","id":"table_id"}],"nextPageToken":"page_token"}"#
            )
        );
    }

    #[tokio::test]
    async fn table_version_response() {
        let response = TableVersionResponse {
            version: TableVersionNumber::new(123),
        };
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get("Delta-Table-Version").unwrap(),
            HeaderValue::from_static("123")
        );
    }
}
