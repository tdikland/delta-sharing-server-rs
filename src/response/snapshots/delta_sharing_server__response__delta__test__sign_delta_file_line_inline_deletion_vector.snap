---
source: src/response/delta.rs
expression: file_line
---
{
  "id": "0d166c1994fcdd317b5f18e2cfab52a0",
  "deletionVectorFileId": null,
  "version": null,
  "timestamp": null,
  "expirationTimestamp": null,
  "deltaSingleAction": {
    "add": {
      "path": "https://bucket/prefix/key.snappy.parquet?sig=foo_signature",
      "partitionValues": {},
      "size": 573,
      "modificationTime": 1619824428000,
      "dataChange": false,
      "stats": null,
      "tags": {},
      "deletionVector": {
        "storageType": "i",
        "pathOrInlineDv": "wi5b=000010000siXQKl0rr91000f55c8Xg0@@D72lkbi5=-{L",
        "offset": null,
        "sizeInBytes": 40,
        "cardinality": 6
      },
      "baseRowId": null,
      "defaultRowCommitVersion": null,
      "clusteringProvider": null
    }
  }
}
